packages/core/src/caches/base-cache.ts

```diff
@@ -0,0 +1,177 @@
+   *
+   * @param run The function to wrap.
+   * @param types An array of {@link CacheKeyConfig}.
+   */
+  mutate<Args extends unknown[], Return>(
+    run: (...args: Args) => Promise<Return>,
+    ...types: Array<CacheKeyConfig<Args, CacheKeyOf<CacheMapT>>>
+  ) {
+    // Intended. We're going to use `this` cache inside another closure.
+    // eslint-disable-next-line @typescript-eslint/no-this-alias, unicorn/no-this-assignment
+    const kvCache = this;
+
+    const mutated = async function (this: unknown, ...args: Args): Promise<Return> {
+      const value = await run.apply(this, args);
+
+      // We don't leverage `finally` here since we want to ensure cache deleting
+      // only happens when the original function executed successfully
+      void Promise.all(
+        types.map(async ([type, cacheKey]) =>
+          trySafe(kvCache.delete(type, cacheKey?.(...args) ?? BaseCache.defaultKey))
+        )
+      );
+
+      return value;
+    };
+
+    return mutated;
+  }
+
+  /**
+   * [Memoize](https://en.wikipedia.org/wiki/Memoization) a function and cache the result. The function execution
+   * will be also cached, which means there will be only one execution at a time.
+   *
+   * @param run The function to memoize.
+   * @param config The object to determine how cache key will be built. See {@link CacheKeyConfig} for details.
+   * @param getExpiresIn A function to determine how long the cache will be expired. The function will be called
+   * with the resolved value from the original function. The return value should be the expire time in seconds.
+   */
+  memoize<
+    Type extends CacheKeyOf<CacheMapT>,
+    Args extends unknown[],
+    Value extends Readonly<CacheMapT[Type]>,
+  >(
+    run: (...args: Args) => Promise<Value>,
+    [type, cacheKey]: CacheKeyConfig<Args, Type>,
+    getExpiresIn?: (value: Value) => number
+  ) {
+    const promiseCache = new Map<unknown, Promise<Readonly<CacheMapT[Type]>>>();
+    // Intended. We're going to use `this` cache inside another closure.
+    // eslint-disable-next-line @typescript-eslint/no-this-alias, unicorn/no-this-assignment
+    const kvCache = this;
+
+    const memoized = async function (
+      this: unknown,
+      ...args: Args
+    ): Promise<Readonly<CacheMapT[Type]>> {
+      const promiseKey = cacheKey?.(...args) ?? BaseCache.defaultKey;
+      const cachedPromise = promiseCache.get(promiseKey);
+
+      if (cachedPromise) {
+        return cachedPromise;
+      }
+
+      const promise = (async () => {
+        try {
+          // Wrap with `trySafe()` here to ignore Redis errors
+          const cachedValue = await trySafe(kvCache.get(type, promiseKey));
+
+          if (cachedValue) {
+            cacheConsole.info(`${kvCache.name} cache hit for', type, promiseKey`);
+            return cachedValue;
+          }
+
+          const value = await run.apply(this, args);
+
+          await trySafe(kvCache.set(type, promiseKey, value, getExpiresIn?.(value)));
```

Comment:
nit: Should we log or report the redis errors, for debugging purpose?

Reply:
We have a global RedisClient onError listener. It will catch the error info and report to the appInsights.